From 5283d40e027ae272f239d4314b776412f73212da Mon Sep 17 00:00:00 2001
From: JW Wang <wangchun@google.com>
Date: Fri, 20 May 2022 19:16:11 +0530
Subject: [PATCH] From: JW Wang <wangchun@google.com> Date: Thu, 31 Mar 2022
 10:06:06 +0800 Subject: [PATCH] Fix NPE

NPE happens when there is an orphaned session which we've
tried to prevent in all cases.

Log an error message if this situation happens.

Bug: 227342978
Test: atest CtsRootPackageInstallerHostTestCases
Change-Id: Ia21323926bd9db1a6f05461904deb45b4c3dd0bc
(cherry picked from commit 07e31dfb1efabc8110d64819f26a06e12a35e020)
Merged-In: Ia21323926bd9db1a6f05461904deb45b4c3dd0bc
(cherry picked from commit e58049a3ea2c056b999c281c7031f9e16e42f809)
Merged-In: Ia21323926bd9db1a6f05461904deb45b4c3dd0bc
---
 .../java/com/android/server/pm/PackageInstallerService.java  | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/services/core/java/com/android/server/pm/PackageInstallerService.java b/services/core/java/com/android/server/pm/PackageInstallerService.java
index ec7a695736b4..53db1e44c3a7 100644
--- a/services/core/java/com/android/server/pm/PackageInstallerService.java
+++ b/services/core/java/com/android/server/pm/PackageInstallerService.java
@@ -347,7 +347,10 @@ public class PackageInstallerService extends IPackageInstaller.Stub implements
                     // Their staging dirs will be removed too                    
                     PackageInstallerSession root = !session.hasParentSessionId()
                            ? session : mSessions.get(session.getParentSessionId());
-                    if (!root.isDestroyed()) {
+                    if (root == null) {
+                        Slog.e(TAG, "freeStageDirs: found an orphaned session: "
+                                + session.sessionId + " parent=" + session.getParentSessionId());
+                    } else if (!root.isDestroyed()) {   
                         root.abandon();
                     }
                 } else {        
-- 
2.17.1

